### PR TITLE
feat(#577): add draft implementation for listing certs.

### LINKED ISSUE #577: [FEATURE] Wallet: listCertificates

### PR DESCRIPTION
**Pull request by mgosek-4chain**
## Linked Issues / Tickets

 "Closes #577".

## Testing Procedure

Describe the tests you've added or any testing steps you've taken.

- [X] I have added new unit tests
- [X] All tests pass locally
- [X] I have tested manually in my local environment

## Checklist:

- [X] I have performed a self-review of my own code
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have run the linter

### CODE REVIEW

File: pkg/wdk/certificate.go [hunk 1]
```diff
@@ -19,10 +92,80 @@ type ListCertificatesResult struct {
...
+// IsEmpty checks whether the VerifierString is empty.
+// Returns true if the string has length 0, otherwise false.
+func (s VerifierString) IsEmpty() bool {
+	return len(s) == 0
+}
+
+// parseSignature converts a HexString into an EC signature.
+// The input `s` is expected to be a concatenation of the R and S values in hex format (64 characters each).
+// Returns a pointer to an ec.Signature containing the parsed R and S values.
+func parseSignature(s primitives.HexString) *ec.Signature {
```
[the remark this answers is not in the document]
Comment (chris-4chain) [reply pkg/wdk/certificate.go]: Agree with copilot, I think this should be checked.

File: pkg/wdk/table_certificate_field.go [hunk 2]
```diff
@@ -16,3 +16,23 @@ type TableCertificateField struct {
...
+// and the corresponding master key from `keyringForSubject`.
+func ParseToTableCertificateFieldSlice(userID int, fields map[string]string, keyringForSubject map[string]string) []*TableCertificateField {
+	tableCertificateFields := make([]*TableCertificateField, 0, len(fields))
+	for k, v := range fields {
+		tableCertificateFields = append(tableCertificateFields, &TableCertificateField{
+			CreatedAt:  time.Now(),
+			UserID:     userID,
+			FieldName:  k,
+			FieldValue: v,
+			MasterKey:  primitives.Base64String(keyringForSubject[k]),
```
Comment (chris-4chain) [inline pkg/wdk/table_certificate_field.go]: If the `keyringForSubject` map does not contain the key `k`, this will silently set the MasterKey to an empty string, which may or may not be the intended behavior.

File: pkg/wallet/internal/testabilities/certificates.go [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+func CreateSampleAcquireCertificateArgs(t *testing.T) wallet.AcquireCertificateArgs {
+	t.Helper()
+	var (
+		certType  wallet.CertificateType
+		serialNum wallet.SerialNumber
+	)
+
+	copy(certType[:], []byte("testType"))
+	copy(serialNum[:], []byte("serialXYZ"))
```
Comment (chris-4chain) [inline pkg/wallet/internal/testabilities/certificates.go]: `[]byte` is unnecessary. Use just:
```go
	copy(serialNum[:], "serialXYZ")
```

File: pkg/wallet/wallet.go [hunk 4]
```diff
@@ -472,16 +475,113 @@ func (w *Wallet) RevealSpecificKeyLinkage(ctx context.Context, args sdk.RevealSp
...
+		Types:      types,
+		Limit:      primitives.PositiveIntegerDefault10Max10000(to.Value(args.Limit)),
+		Offset:     primitives.PositiveInteger(to.Value(args.Offset)),
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to list certificates with given list certificates args: %w", err)
+	}
+
+	certs := make([]sdk.CertificateResult, 0, len(listCertificatesResult.Certificates))
+	for _, storedModel := range listCertificatesResult.Certificates {
```
Comment (chris-4chain) [inline pkg/wallet/wallet.go:576]: The code here could be in the mapping.

File: pkg/wallet/wallet.go [hunk 5]
```diff
@@ -472,16 +475,113 @@ func (w *Wallet) RevealSpecificKeyLinkage(ctx context.Context, args sdk.RevealSp
...
+		RevocationOutpoint: args.RevocationOutpoint,
+		Fields:             args.Fields,
+		Signature:          args.Signature,
+	}
+
+	return &cert, nil
 }
 
 // ListCertificates lists identity certificates belonging to the user, filtered by certifier(s) and type(s).
 func (w *Wallet) ListCertificates(ctx context.Context, args sdk.ListCertificatesArgs, originator string) (*sdk.ListCertificatesResult, error) {
```
Comment (chris-4chain) [inline pkg/wallet/wallet.go:561]: The fields `Privileged` and `PrivilegedReason` are not used from the `args` - is it ok?

File: pkg/wallet/wallet_acquire_certificate_test.go [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+import (
+	"github.com/bsv-blockchain/go-sdk/wallet"
+	"github.com/bsv-blockchain/go-wallet-toolbox/pkg/internal/fixtures"
+	"github.com/bsv-blockchain/go-wallet-toolbox/pkg/wallet/internal/testabilities"
+	"github.com/go-softwarelab/common/pkg/to"
+	"github.com/stretchr/testify/require"
+)
+
+func (s *WalletTestSuite) Test_AcquireCertificate() {
```
Comment (chris-4chain) [inline pkg/wallet/wallet_acquire_certificate_test.go:10]: There are no error/edge cases covered. Please add tests that checks if this works properly with some empty/wrong data as args

File: pkg/wallet/wallet_list_certificates_test.go [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+import (
+	sdkprimitives "github.com/bsv-blockchain/go-sdk/primitives/ec"
+	"github.com/bsv-blockchain/go-sdk/wallet"
+	"github.com/bsv-blockchain/go-wallet-toolbox/pkg/internal/fixtures"
+	"github.com/bsv-blockchain/go-wallet-toolbox/pkg/wallet/internal/testabilities"
+	"github.com/stretchr/testify/require"
+)
+
+func (s *WalletTestSuite) Test_ListCertificates() {
```
Comment (chris-4chain) [inline pkg/wallet/wallet_list_certificates_test.go:15]: Similarly, please add a few corner cases. Additionally, I would test what happens when there is an empty result.

File: pkg/wdk/certificate.go [hunk 8]
```diff
@@ -1,16 +1,89 @@
...
+// WalletCertificateFieldMap represents a mapping of wallet certificate fields.
+// The keys are constrained to primitives.StringUnder50Bytes (strings up to 50 bytes),
+// and the values are regular strings.
+type WalletCertificateFieldMap map[primitives.StringUnder50Bytes]string
+
+// ToMap converts the WalletCertificateFieldMap to a standard map[string]string.
+// This is useful for interoperability with functions or APIs that expect regular string maps.
+// Returns a new map where each key and value is converted to a plain string.
+func (m WalletCertificateFieldMap) ToMap() map[string]string {
+	out := make(map[string]string)
```
Comment (chris-4chain) [inline pkg/wdk/certificate.go]:
```suggestion
	out := make(map[string]string, len(m))
```

File: pkg/wdk/certificate.go [hunk 9]
```diff
@@ -19,10 +92,80 @@ type ListCertificatesResult struct {
...
+// Returns true if the map has no elements, otherwise false.
+func (m KeyringMap) IsEmpty() bool {
+	return len(m) == 0
+}
+
+// ToMap converts the KeyringMap to a standard map[string]string.
+// This is useful for interoperability with functions or APIs that expect regular string maps.
+// Returns a new map where both keys and values are converted to plain strings.
+func (m KeyringMap) ToMap() map[string]string {
+	out := make(map[string]string)
```
Comment (chris-4chain) [inline pkg/wdk/certificate.go]:
```suggestion
	out := make(map[string]string, len(m))
```

File: pkg/wdk/certificate.go [hunk 10]
```diff
@@ -19,10 +92,80 @@ type ListCertificatesResult struct {
...
+// parseCertificationType decodes a base64-encoded string into an sdk.CertificateType.
+// Returns the decoded CertificateType, or an error if decoding fails.
+func parseCertificationType(s string) (sdk.CertificateType, error) {
+	bytes, err := base64.StdEncoding.DecodeString(s)
+	if err != nil {
+		return sdk.CertificateType{}, fmt.Errorf("failed to decode certificate type: %w", err)
+	}
+
+	var certType sdk.CertificateType
+	copy(certType[:], bytes)
```
Comment (chris-4chain) [inline pkg/wdk/certificate.go]: Shouldn't you check if returned `bytes` have correct length? Like:
```go
if len(bytes) != len(certType) {
    return fmt.Errorf("some error")
}

```

File: pkg/wdk/table_certificate_field.go [hunk 11]
```diff
@@ -16,3 +16,23 @@ type TableCertificateField struct {
...
+type TableCertificateFieldSlice []TableCertificateField
+
+// ParseToTableCertificateFieldSlice converts a map of certificate fields into a slice of TableCertificateField pointers.
+// Each entry in the `fields` map becomes a TableCertificateField, populated with the user ID, field name, field value,
+// and the corresponding master key from `keyringForSubject`.
+func ParseToTableCertificateFieldSlice(userID int, fields map[string]string, keyringForSubject map[string]string) []*TableCertificateField {
+	tableCertificateFields := make([]*TableCertificateField, 0, len(fields))
+	for k, v := range fields {
+		tableCertificateFields = append(tableCertificateFields, &TableCertificateField{
+			CreatedAt:  time.Now(),
```
Comment (chris-4chain) [inline pkg/wdk/table_certificate_field.go]: I think that the `CreatedAt` shouldn't be set here. It will be set by storage automatically. Besides, look how this is implemented on the storage level - the `CreatedAt` is not included there (during the mapping).

File: pkg/wallet/wallet.go (hunk 5 above)
Comment (mgosek-4chain) [reply pkg/wallet/wallet.go:561]: Arguments usage has been added.

File: pkg/wdk/certificate.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (mgosek-4chain) [reply pkg/wdk/certificate.go]: Done :)

File: pkg/wdk/table_certificate_field.go (hunk 2 above)
Comment (mgosek-4chain) [reply pkg/wdk/table_certificate_field.go]: Fixed :)

File: pkg/wallet/wallet_acquire_certificate_test.go (hunk 6 above)
Comment (mgosek-4chain) [reply pkg/wallet/wallet_acquire_certificate_test.go:10]: Done :)

File: pkg/storage/mappings.go [hunk 12]
```diff
@@ -34,8 +34,8 @@ func certModelToResult(model *entity.Certificate) *wdk.CertificateResult {
 	}
 }
 
-func certificateModelFieldsToKeyringResult(fields []entity.CertificateField) map[primitives.StringUnder50Bytes]primitives.Base64String {
-	result := make(map[primitives.StringUnder50Bytes]primitives.Base64String, len(fields))
+func certificateModelFieldsToKeyringResult(fields []entity.CertificateField) wdk.KeyringMap {
```
Comment (chris-4chain) [inline pkg/storage/mappings.go]: This function blindly converts field.FieldValue to primitives.Base64String without asserting that field.FieldValue is base64-encoded. If the underlying data is not guaranteed to be base64, this can lead to errors downstream when decoding.

File: pkg/wdk/certificate.go [hunk 13]
```diff
@@ -19,10 +98,95 @@ type ListCertificatesResult struct {
...
+	serialBytes, err := base64.StdEncoding.DecodeString(s)
+	if err != nil {
+		return sdk.SerialNumber{}, fmt.Errorf("failed to decode certificate serial number: %w", err)
+	}
+
+	var serial sdk.SerialNumber
+	copy(serial[:], serialBytes)
+
+	if len(serialBytes) != len(serial) {
+		return sdk.SerialNumber{}, fmt.Errorf("serial bytes length: %d is not equal to sdk.SerialNumber bytes length: %d", len(serialBytes), len(serial))
```
Comment (chris-4chain) [inline pkg/wdk/certificate.go:168]: I would rather make this check "before" the copy phase.
The way I see it:
- The serial is a fixed size array (`32`) so if the length of `serialBytes` is not 32 - there is no point to proceed and we should return an error right away.

**Review by chris-4chain — APPROVED** [review]
